### ISSUE TITLE
Reified existential equality constraints are different between GHC 7.10 and 8.0

### ISSUE THREAD

**Issue by RyanGlScott** [glguy/th-abstraction#7]
Compare this GHCi 7.10.3 session:

```
$ /opt/ghc/7.10.3/bin/ghci
GHCi, version 7.10.3: http://www.haskell.org/ghc/  :? for help
λ> :set -XGADTs -XTemplateHaskell 
λ> :m + Language.Haskell.TH Language.Haskell.TH.Datatype 
λ> data Foo a = a ~ Int => MkFoo
λ> putStrLn $(reify ''Foo >>= stringE . show)
TyConI (DataD [] Ghci1.Foo [KindedTV a_1627410848 StarT] [ForallC [] [AppT (AppT EqualityT (VarT a_1627410848)) (ConT GHC.Types.Int)] (NormalC Ghci1.MkFoo [])] [])
λ> putStrLn $(reifyDatatype ''Foo >>= stringE . show)
DatatypeInfo {datatypeContext = [], datatypeName = Ghci1.Foo, datatypeVars = [SigT (VarT a_1627410848) StarT], datatypeVariant = Datatype, datatypeCons = [ConstructorInfo {constructorName = Ghci1.MkFoo, constructorVars = [], constructorContext = [AppT (AppT EqualityT (VarT a_1627410848)) (ConT GHC.Types.Int)], constructorFields = [], constructorVariant = NormalConstructor}]}
```

to this GHC 8.0.2 one:

```
$ /opt/ghc/8.0.2/bin/ghci
GHCi, version 8.0.2: http://www.haskell.org/ghc/  :? for help
Loaded GHCi configuration from /home/rgscott/.ghci
λ> :set -XGADTs -XTemplateHaskell 
λ> :m + Language.Haskell.TH Language.Haskell.TH.Datatype 
λ> data Foo a = a ~ Int => MkFoo
λ> putStrLn $(reify ''Foo >>= stringE . show)
TyConI (DataD [] Ghci1.Foo [KindedTV a_6989586621679042262 StarT] Nothing [ForallC [KindedTV a_6989586621679042262 StarT] [AppT (AppT (ConT Data.Type.Equality.~) (VarT a_6989586621679042262)) (ConT GHC.Types.Int)] (GadtC [Ghci1.MkFoo] [] (AppT (ConT Ghci1.Foo) (VarT a_6989586621679042262)))] [])
λ> putStrLn $(reifyDatatype ''Foo >>= stringE . show)
DatatypeInfo {datatypeContext = [], datatypeName = Ghci1.Foo, datatypeVars = [SigT (VarT a_6989586621679042262) StarT], datatypeVariant = Datatype, datatypeCons = [ConstructorInfo {constructorName = Ghci1.MkFoo, constructorVars = [], constructorContext = [AppT (AppT (ConT Data.Type.Equality.~) (VarT a_6989586621679042262)) (ConT GHC.Types.Int)], constructorFields = [], constructorVariant = NormalConstructor}]}
```

Namely, there's a difference in the constructor's context: we have `EqualityT` in 7.10.3, but `ConT Data.Type.Equality.~` in 8.0.2. Perhaps it's worth normalizing the latter to `EqualityT`?

**Comment by glguy** [comment]
Normalizing seems reasonable to me

**Comment by glguy** [comment]
It won't be possible to make this uniform.

Before 7.10.3 there was the separate Pred type with EqualP, GHC 7.10.3 uses EqualityT, after use ConT ''(~)

**Comment by glguy** [comment]
https://travis-ci.org/glguy/th-abstraction/builds/235890057

**Comment by glguy** [comment]
Perhaps the solution is to provide a predicate: `isEqualityT :: Pred -> Bool` or `fromEqualityT :: Pred -> Maybe (Type, Type)`

**Comment by RyanGlScott** [comment]
I think `fromEqualityT` would be a good solution, then.

**Comment by glguy** [comment]
Addressed in 4c74a32d4fd824fb238ab34812f4a846b03d6d20

**Comment by RyanGlScott** [comment]
Looks good!
